### DISCUSSION TITLE
Plugin scaffolding tool

### DISCUSSION THREAD

**Discussion by leventebalogh** [grafana/grafana#40568] [Plugins]
| **Author(s)** | @jackw @mckn @leventebalogh  |
| :---- | :--- |
| **Created**     | 2020-09-21                                                                                                                                                 |
| **Status**      | **Work in progress**                                                                                                                                       |
| **Approver(s)** | @torkelo @dprokop @ryantxu @marcusolsson                                                                                                                                                            |
| **Informed**    |  @wbrowne @marefr @grafana/cloud-datasources                                                                                                                                                           |
| **POC**         | https://github.com/grafana/create-plugin    <https://github.com/grafana/grafana/issues/39824>                                                            | 

# ToC

[Background](https://docs.google.com/document/d/19PX9ngg9n0qKrWoPZjE_y5xjOwK-ePtr92cOQxMqRwM/edit#heading=h.ezbem4jbn28v)
[Problems](https://docs.google.com/document/d/19PX9ngg9n0qKrWoPZjE_y5xjOwK-ePtr92cOQxMqRwM/edit#heading=h.hacp0w15ubvu)
[Goals](https://docs.google.com/document/d/19PX9ngg9n0qKrWoPZjE_y5xjOwK-ePtr92cOQxMqRwM/edit#heading=h.up4sjwu3my9v)
[Nice to have](https://docs.google.com/document/d/19PX9ngg9n0qKrWoPZjE_y5xjOwK-ePtr92cOQxMqRwM/edit#heading=h.9xsagwfx03gt)
[Proposals](https://docs.google.com/document/d/19PX9ngg9n0qKrWoPZjE_y5xjOwK-ePtr92cOQxMqRwM/edit#heading=h.e1oa33xbl767)
[Questions](https://docs.google.com/document/d/19PX9ngg9n0qKrWoPZjE_y5xjOwK-ePtr92cOQxMqRwM/edit#heading=h.vdn5clilbq1i)
[Consensus](https://docs.google.com/document/d/19PX9ngg9n0qKrWoPZjE_y5xjOwK-ePtr92cOQxMqRwM/edit#heading=h.ve4du26wqsib)

# Background

Currently we recommend the plugin authors to use `@grafana/toolkit` to scaffold, build, test and run plugins during development. It is a tool that is used not only by plugin developers but also in the following scenarios (grouped by area):

**Current responsibilities of `@grafana/toolkit`:**

**Plugins**

- Build plugins.
- Bundle plugins.
- Test plugins.
- Scaffold plugins.
- Scaffold parts of a plugin.
- Sign plugins
- CI build plugins (marked deprecated for circle-ci)
- Create and upload a Github release for a plugin.
- Updated the circle CI configuration for a plugin.

**Grafana**

- Build Grafana core.
- Build and package the `@grafana/*`-packages.
- Build the grafana/toolkit.
- Importing test search data into Grafana.
- Builds bundled plugins.

**Release Grafana**

- Export changelog while doing a Grafana release.
- Export list of commits to cherry-pick while doing a Grafan release.
- Close the current released milestone.

# Problems

In our opinion the main problems with the current tooling are following:  
  
- **No Windows support** - the current version of the `@grafana/toolkit` does not run well on Windows. Some of the commands will fail due to expecting a unix os. This is mostly due to the `execa` library.
- **Hard to customize** - the scaffolded project will use `@grafana/toolkit` to build, run, and test the project, which hides the build logic and makes it super hard to debug or customize. (In the current way you would need to go to the source code of `@grafana/toolkit` to understand what is going on).
- **Does too many things** - the current version of the `@grafana/toolkit` does too many different things, which makes it hard to extend and maintain.
- **Has a lot of dependencies** - due to the number of use cases it needs to cover it has a lot of external dependencies which makes it slow for being used as a one-off tool.
- **Has frontend prerequisites** - the current tooling requires having `npm` / `npx` installed and having at least some basic knowledge of the frontend ecosystem.
- **Out of sync starter templates** - Currently there is no understanding of the`@grafana/toolkit` version or the starter kit that is downloaded. We assume that the template will always be compatible with the latest `@grafana/toolkit` version however this isn’t always the case. Changes to the starter template can lead to devs opening issues about `@grafana/toolkit` not working correctly.

# Goals

We would like a tool that covers the following main features:

- **Support all major platforms** - works on Windows, Linux and macOS
- **Fast and simple** - scaffolding is as easy as running a simple command and is fast
- **Easy to customize** - scaffolds a plugin template that is easy to debug and customize
- **Offline support** - the main workflow can be used while coding on an airplane
- **Support different Grafana versions** - can be used to scaffold plugins for different major Grafana versions
- **Easy-to-access executable** - the tool should be distributable to multiple channels (`npm`, `brew`, `zip`) so it is easy for both frontend and backend developers to set it up and use it.
- **Examples that are in sync with the templates** - there are always going to be scenarios that are not that simple and thus an example can give guidance to the developers.

# Nice to have

- **Tested templates and examples** - we would also like to introduce a set of tests that can be run to verify the functionality of the generated starter template/examples.

# Proposals

Both of our proposals are based on the idea that we introduce a new single purpose tooling that should live outside of the main grafana repository. The benefit of this is that we can release, publish and manage its dependencies in its own release cycle.

_Edit 2021-10-19 12:03:_
One advantage of introducing a new tool would be that we can deprecate those commands in grafana-toolkit in favor of this new tool without breaking any existing plugins that are currently using them.
_End Edit_

We would also like to replace the usage of `@grafana/toolkit` to build/test/run the scaffolded plugins. Instead we are suggesting that we generate the underlying webpack commands and configuration. This will make it more transparent to the plugin developers and make it easier to extend and debug the build/test/run commands.

If something fails they have the whole webpack community that can help them to figure it out instead of needing to rely on Grafana to debug it in their setup.

## Proposal 1: _have a single repository_

- Have a single repository which holds

  - The scaffolding tool’s source and binary
  - The different templates that we would like to be able to scaffold
  - The different examples that would be used for reference and testing.

- [Idea of accessing the templates when running the tool](https://docs.google.com/document/d/19PX9ngg9n0qKrWoPZjE_y5xjOwK-ePtr92cOQxMqRwM/edit#heading=h.6bj2846yls2i)

| Pros                                                                                           	| Cons                                                                               	|
|------------------------------------------------------------------------------------------------	|------------------------------------------------------------------------------------	|
| As we can have common templates for the different types of plugins, it can be easier to manage 	| We cannot reuse the existing starter repos                                         	|
| Probably easier to run tests / checks against the templates before releasing anything new      	| We need to update the documentation referencing the starter template repositories. 	|
| A simpler structure which is easier to oversee                                                 	|                                                                                    	|
| Adding a new type of plugin is easier                                                          	|                                                                                    	|

## Proposal 2: _have multiple repositories_

- Have a repository that holds the scaffolding tool
- Have different repositories for each starter template (like how it is currently)
- Have different repositories for each example
- [Idea of accessing the templates when running the tool](https://docs.google.com/document/d/19PX9ngg9n0qKrWoPZjE_y5xjOwK-ePtr92cOQxMqRwM/edit#heading=h.6bj2846yls2i)

| Pros                                                                                  	| Cons                                                                                              	|
|---------------------------------------------------------------------------------------	|---------------------------------------------------------------------------------------------------	|
| We can keep the repos that we currently have for the starter templates                	| Updating something needs to open PRs in multiple repos                                            	|
| The templates can be modified (fixed) without the need to update the scaffolding tool 	| Adding a new plugin type involves creating a new repository and also to edit the scaffolding tool 	|
| We don’t need to update the documentation that is referencing those repositories.     	| It’s harder to keep the scaffolding tool in sync with the templates                               	|

# Questions

## How to access the templates?

There are two ways to access the templates from the scaffolding tool:

### Bundle them with the binary _(preferred)_

This means that every version of the scaffolding tool is going to have the templates bundled with the same version. If we publish a fix, the users are going to need to update their version of the scaffolding tool in order to have that fix in the templates. _(This should be less of a problem by using `npx` as it is always running the latest version of the binary.)_  

**We are preferring this way because on-prem users who might not have access to the world-wide-web can use the tool as well.**

### Fetch them on each run

This means that every time the scaffolding tool is run the latest version of the particular template is fetched. Although this would make it easier to update templates without the user needing to update the scaffolding tool itself, it can block on-prem users who don’t have direct access to the world-wide-web from accessing these templates.

## Support for nested plugins

Currently we have decided not to support and encourage nested plugins with the new scaffolder tool. We don’t think that they are strongly justified, and there are better alternatives to solve these problems.

# Consensus

Currently we have decided to go with **_Proposal 1_** and create a [PoC](https://github.com/grafana/create-plugin) for it.

**Comment by torkelo** [comment]
Awesome issue and goals/proposals! A major overhaul of the toolkit really needed (especially the CI bits). 

> Easy-to-access executable - the tool should be distributable to multiple channels (npm, brew, zip) 

Not sure the value vs effort here is great. An npm package is one of the most universal and easiest to install out there, having zip & brew packages will not make much of a difference I think. 

> Single repo & multi repo

Think a single repo with tool & templates would probably make things easier. And having templates part of the tool package will make sure they do not get out of sync.

>  Instead we are suggesting that we generate the underlying webpack commands and configuration. This will make it more transparent to the plugin developers and make it easier to extend and debug the build/test/run commands.

This is probably the best solution, question is when we want to make webpack config or jest config change all plugin developers will need to apply that manually. Unless we have some configuration inheritance scheme, but that could be complex/error prone (overrides, etc). 

> Offline support - the main workflow can be used while coding on an airplane 

This seems like less of an issue, the current toolkit works great on an airplane (for building & testing), sure creating a new plugin is an issue, but after the create you would need to run npm/yarn install and that would fail if you have no internet. So spending a lot of work on offline support seems not worth it, think better to focus on the issue of things being out of sync due to fetched at runtime rather than being bundled

**Reply by mckn** [reply]
Awesome feedback! Thank you 🙏🏻 

> > Easy-to-access executable - the tool should be distributable to multiple channels (npm, brew, zip)
> 
> Not sure the value vs effort here is great. An npm package is one of the most universal and easiest to install out there, having zip & brew packages will not make much of a difference I think.

When we added this we had the following mindset. Let's start to publish it to npm and see if that is enough. If we get requests on providing it via more channels we can always do that. So the idea here was not to pick a stack that would prevent us from doing that. If that makes sense.

> > Instead we are suggesting that we generate the underlying webpack commands and configuration. This will make it more transparent to the plugin developers and make it easier to extend and debug the build/test/run commands.
> 
> This is probably the best solution, question is when we want to make webpack config or jest config change all plugin developers will need to apply that manually. Unless we have some configuration inheritance scheme, but that could be complex/error prone (overrides, etc).

We have multiple ideas of how to handle this but the most simple one would be to provide migration steps in the same way as we would do for the rest of the code/APIs. If we feel that we could automate these migration steps in some way that might be an option in the future but let's start simple. 

I'm not a huge fan of the override or inheritance scheme idea since it will add complexity and we might end up where we are today. But we are opening this discussion to get feedback on these things so would love to hear what others think. 

> > Offline support - the main workflow can be used while coding on an airplane
> 
> This seems like less of an issue, the current toolkit works great on an airplane (for building & testing), sure creating a new plugin is an issue, but after the create you would need to run npm/yarn install and that would fail if you have no internet. So spending a lot of work on offline support seems not worth it, think better to focus on the issue of things being out of sync due to fetched at runtime rather than being bundled

Yes, totally agree! We added this when talking about how to handle templates for multiple Grafana versions. So our idea was to include the latest version of the plugin templates together with the tool. If you would like to scaffold a plugin for an older version of Grafana we would fetch that from github. So this was brought up as a plus with the approach described above.

**Comment by dprokop** [comment]
Great initiative folks 🚀 

Few cents from me:

> Hard to customize - the scaffolded project will use @grafana/toolkit to build, run, and test the project, which hides the build logic and makes it super hard to debug or customize. (In the current way you would need to go to the source code of @grafana/toolkit to understand what is going on).
>We would also like to replace the usage of @grafana/toolkit to build/test/run the scaffolded plugins. Instead we are suggesting that we generate the underlying webpack commands and configuration. This will make it more transparent to the plugin developers and make it easier to extend and debug the build/test/run commands.

Keep in mind this is on purpose. When we created toolkit the situation was that every single plugin out there had it's own custom webpack(or whatever) config. The goal was to provide a standard configs that would suit **the majority of the needs** while at the same time providing means to extend the configuration. Anyone can extend the webpack config currently with https://github.com/grafana/grafana/blob/1d689888b0fc2de2dbed6e606eee19561a3ef006/packages/grafana-toolkit/src/config/webpack.plugin.config.ts. I'm not sure if it's documented tho :/ 
Similar thing applies to jest configuration, tsconfig and prettier are  on the other hand customizable with a std approach (json/rc files). 

IMO extending webpack config should be relatively hard as it can create an incentive to replace the config completely.

So, the customization is hard on purpose, yet possible. Maybe a guide on how to customize configs would be good. 

The mode that you are proposing here is similar to CRA's `eject` which I would expect to be an option, not a default.

I think another consequence of giving the full control of the configs to the consumers is that we completely lose control of the pipeline. No way to run std commands in the CI so pretty much we cannot have reliable CI pipeline for ANY plugin.

> Easy-to-access executable - the tool should be distributable to multiple channels (npm, brew, zip)

I agree with @torkelo on this one. I don't think this will bring that much value. Is this something that's been raised as issue by the Comunity?

>Does too many things - the current version of the @grafana/toolkit does too many different things, which makes it hard to extend and maintain.

Yes yes yes - as long as toolkit serves it's original purpose (scaffold + dev process) we will be home I think.

Regarding the templates - I would opt for having them bundled with toolkit. The current approach showed multiple times how unmaintained the templates are. So even with the cost of new toolkit release required with template updates I think we will be in a better situation that currently.

**Reply by leventebalogh** [reply]
Thanks a lot for your thoughts @dprokop! :pray:

**Easy-to-access executable**
I agree with you and @torkelo on this one, we mostly added it as a possible benefit just to lower the bar for entry, but probably not something that would worth putting a lot of effort into at this point. (as @mckn said we can keep paying attention on this and add it later if it is coming up as a serious need from the community)

**Bundling templates together with the tool**
Amazing, it seems like it's forming to be the way forward.

**Extendable build configuration**
Thanks for giving us more explanation on this!

I think all the things that you listed are valid points, we were also thinking about these a lot. Both approaches seem to have their pros and cons, I will try to explain why we have chosen ejecting the configs by default in the end.

- _would make the tool less complex_ (as it wouldn't need to be concerned about where the project lives and how to merge the custom configs with the default ones)
- _would be simpler to develop/maintain the tool itself_  - there would be no need to [link](https://github.com/grafana/grafana/blob/1c64f7d3ffc9a87d085830544ded77ab7e58263f/packages/grafana-toolkit/src/cli/tasks/plugin.dev.ts#L11) the tool with yarn (as the package itself is not used for running the commands)
- _would require to document less things for the users_ - they wouldn't need to learn about how they can customise any of the configs in the projects, the files as a single source of truth would be in the repo
- _less chance to break plugin workflows_ - although updating the tooling would require a manual step (which is drawback), there would be also less chance to break the workflows of the plugins (in most of the cases the builds keep working if the other grafana packages are updated and can produce a working package) 

> I think another consequence of giving the full control of the configs to the consumers is that we completely lose control of the pipeline. No way to run std commands in the CI so pretty much we cannot have reliable CI pipeline for ANY plugin.

You are right that this wouldn't take us towards having a centralised CI workflow which can test all the plugins.

Although giving the plugin developers all the guidance for building a plugin, we would try focusing on having a clear contract on **_what_** we expect (the structure of the packaged plugin and the available Grafana APIs) and give them more freedom on **_how_** they actually build it. The scaffolded plugin would still have a standard set of commands and working CI pipelines, so most of the plugins would probably still have the same "interface", but customisation would be a bit easier. 

We were also thinking about introducing a new command (e.g. `yarn validate-package`) for the scaffolded plugins which could help checking if the generated package is in a compatible format for the targeted Grafana version(s).

**Reply by dprokop** [reply]
> ... and give them more freedom on how they actually build it. The scaffolded plugin would still have a standard set of commands and working CI pipelines, so most of the plugins would probably still have the same "interface", but customization would be a bit easier.

I think the freedom is there already, just not documented. 

> less chance to break plugin workflows

I'm also thinking about the plugin review process. This will now require configs reviews as well. What's bundled with the plugin, etc. I think this is an overhead that will not make the process any simpler. 

>but customisation would be a bit easier.

But is customization a real issue now? How much customization do we think the user wants? 
I'm in a middle of migrating BQ ds now. Quite a complex one. Originally that data source had plenty of custom webpack configuration. With toolkit I didn't have to know there is webpack under the hood, because the entire dev flow just worked after removing the original webpac config. And I'm more than happy that I don't have to even think there is any webpack config for my plugin.

Out of curiosity: _is the proposed config approach result of users' complains?_

---

Don't get me wrong. I'm all in for the toolkit changes, I'm just concerned that the proposed config changes I think are taking us a step back. With toolkit commands, we have a clear contract on what we expect the user to use. Without it (even with ejected webpack config) we are saying and I also thing encouraging users - do whatever you want. And hence we loose plenty of control over what we expect (apart from the contract, which from what I understand is going to be npm scrips convention).

**Comment by mckn** [comment]
@briangann @grafana/cloud-datasources @srclosson @alexanderzobnin would love to get your feedback on this 🙏🏻

**Reply by leventebalogh** [reply]
Also pinging @besartberisha @domasx2 @bfmatei @darrenjaneczek, as you were working closely with tooling around plugins in the past.

**Reply by marefr** [reply]
Adding @toddtreece @gingerwizard

**Comment by mckn** [comment]
An update on what we currently are working on in regards to this.

The plan is to deprecate all scaffolding of plugin features of the `grafana-toolkit` in favor of a new `create-plugin` tool (https://github.com/grafana/create-plugin). 

The new tool will be able to scaffold app, panel and datasource plugins. It will also be able to support to scaffolding of plugins for multiple versions of Grafana. Our idea is to have one tool that does one job really well. 

We going back and forth regarding the scaffolded plugin. If it will use `grafana-toolkit` to build, run and test the plugin or if we should scaffold a set of default configurations to handle this. Currently we are leaning towards the latter since it, according to us, has more advantages compared to our current solution.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). This looks to be an ongoing discussion, and we'd like those about grafana to go to our[ community forum](https://community.grafana.com/). Thank you!
